File: WebClient SDK/src/default/tiles/krDocStateTileExtension.ts
import { TileExtension, ITileGlobalExtensionContext, Tile, TileGroups, TileEvaluationEventArgs, TileHotkey, ITileLocalExtensionContext } from 'tessa/ui/tiles';
import { UIContext } from 'tessa/ui';
import { DeleteIntegerCardOperation } from 'tessa/ui/cards';
import { createTypedField, DotNetType } from 'tessa/platform';
import { userSession } from 'common/utility';

export class KrDocStateTileExtension extends TileExtension {

  public initializingGlobal(context: ITileGlobalExtensionContext) {
    const panel = context.workspace.leftPanel;

    const viewsOther = panel.tryGetTile('ViewsOther');
    if (viewsOther) {
      const deleteCardFromView = viewsOther.tryGetTile('DeleteCardFromView');

      viewsOther.tiles.push(
        new Tile({
          name: 'DeleteKrDocStateFromView',
          caption: '$UI_Tiles_DeleteCardFromView',
          icon: 'icon-thin-059',
          contextSource: panel.contextSource,
          group: TileGroups.Views,
          order: deleteCardFromView ? deleteCardFromView.order : 100,
          command: KrDocStateTileExtension.deleteKrDocStateFromViewAsync,
          evaluating: KrDocStateTileExtension.deleteKrDocStateFromViewEvaluating,
          toolTip: '$UI_Tiles_DeleteCardFromView_ToolTip'
        })
      );
    }
  }

  public initializingLocal(context: ITileLocalExtensionContext) {
    const panel = context.workspace.leftPanel;
    if (!panel.context.viewContext) {
      return;
    }

    const hotkeyStorage = panel.contextSource.hotkeyStorage;

    const viewsOther = panel.tryGetTile('ViewsOther');
    if (!viewsOther) {
      return;
    }

    const deleteKrDocStateFromView = viewsOther.tryGetTile('DeleteKrDocStateFromView');
    if (deleteKrDocStateFromView) {
      hotkeyStorage.addTileHotkey(new TileHotkey(deleteKrDocStateFromView, 'Ctrl+D', 'KeyD', { ctrl: true }));
      if (userSession.isAdmin) {
        hotkeyStorage.addTileHotkey(
          new TileHotkey(deleteKrDocStateFromView, 'Ctrl+Alt+D', 'KeyD', { ctrl: true, alt: true })
        );
      }
    }
  }

  private static async deleteKrDocStateFromViewAsync() {
    const viewContext = UIContext.current.viewContext;
    if (viewContext) {
      // не включаем режим принудительного удаления без возможности восстановления, т.к. виртуальная карточка
      // и так удаляется без восстановления, но режим принудительного удаления отображает другие предупреждения перед удалением
      const withoutBackupOnly = false;

      const operation = new DeleteIntegerCardOperation(
        (request, item) => request.info['StateID'] = createTypedField(item.cardId, DotNetType.Int32),
        _item => 'e83a230a-f5fc-445e-9b44-7d0140ee69f6', // KrDocStateTypeID
        withoutBackupOnly
      );

      await operation.startAsync(viewContext);
    }

  }

  private static deleteKrDocStateFromViewEvaluating(e: TileEvaluationEventArgs) {
    const viewContext = e.currentTile.context.viewContext;

    e.setIsEnabledWithCollapsing(e.currentTile,
      !!viewContext
      && !!viewContext.view
      && viewContext.view.metadata.alias === 'KrDocStateCards'
    );
  }

}

File: Source/Extensions/Tessa.Extensions.Default.Client/Tiles/KrDocStateTileExtension.cs
using System;
using System.Threading.Tasks;
using Tessa.Cards;
using Tessa.Extensions.Default.Shared;
using Tessa.Extensions.Platform.Client.Tiles;
using Tessa.Platform;
using Tessa.Platform.Runtime;
using Tessa.UI;
using Tessa.UI.Cards;
using Tessa.UI.Tiles;
using Tessa.UI.Tiles.Extensions;
using Tessa.UI.Views;
using Tessa.Views;
using Tessa.Views.Metadata;

namespace Tessa.Extensions.Default.Client.Tiles
{
    public sealed class KrDocStateTileExtension :
        TileExtension
    {
        #region Constructors

        public KrDocStateTileExtension(
            ICardRepository cardRepository,
            ICardMetadata cardMetadata,
            ISession session)
        {
            this.cardRepository = cardRepository;
            this.cardMetadata = cardMetadata;
            this.session = session;
        }

        #endregion

        #region Fields

        private readonly ICardRepository cardRepository;

        private readonly ICardMetadata cardMetadata;

        private readonly ISession session;

        #endregion

        #region Private Methods

        private async void DeleteKrDocStateFromViewAsync(object parameter)
        {
            IViewContext viewContext = UIContext.Current.ViewContext;
            if (viewContext != null)
            {
                // не включаем режим принудительного удаления без возможности восстановления, т.к. виртуальная карточка
                // и так удаляется без восстановления, но режим принудительного удаления отображает другие предупреждения перед удалением
                const bool withoutBackupOnly = false;

                var operation = new DeleteIntegerCardOperation(
                    (request, item) => request.Info[DefaultExtensionHelper.StateIDKey] = (int) item.CardID,
                    async (item, ct) => DefaultCardTypes.KrDocStateTypeID,
                    withoutBackupOnly,
                    this.cardRepository,
                    this.cardMetadata);

                await operation.StartAsync(viewContext);
            }
        }


        private static async void DeleteKrDocStateFromViewEvaluating(object sender, TileEvaluationEventArgs e)
        {
            IViewContext viewContext = e.CurrentTile.Context.ViewContext;

            var deferral = e.Defer();
            try
            {
                bool isEnabled = viewContext != null;
                if (isEnabled)
                {
                    ITessaView view = viewContext.View;
                    IViewMetadata metadata = view != null ? await view.GetMetadataAsync() : null;

                    isEnabled = metadata?.Alias == DefaultViewAliases.KrDocStateCards;
                }

                e.SetIsEnabledWithCollapsing(e.CurrentTile, isEnabled);
            }
            catch (Exception ex)
            {
                deferral.SetException(ex);
            }
            finally
            {
                deferral.Dispose();
            }
        }

        #endregion

        #region Base Overrides

        public override Task InitializingGlobal(ITileGlobalExtensionContext context)
        {
            ITilePanel panel = context.Workspace.LeftPanel;

            ITile viewsOther = panel.Tiles.TryGet(TileNames.ViewsOther);
            if (viewsOther != null)
            {
                ITile deleteCardFromView = viewsOther.Tiles.TryGet(TileNames.DeleteCardFromView);

                viewsOther.Tiles.Add(
                    new Tile(
                        DefaultTileNames.DeleteKrDocStateFromView,
                        "$UI_Tiles_DeleteCardFromView",
                        context.Icons.Get("Thin59"),
                        panel,
                        new DelegateCommand(this.DeleteKrDocStateFromViewAsync),
                        TileGroups.Views,
                        deleteCardFromView?.Order ?? 100,
                        toolTip: TileHelper.GetToolTip("$UI_Tiles_DeleteCardFromView_ToolTip", TileKeys.DeleteCardFromView),
                        evaluating: DeleteKrDocStateFromViewEvaluating));
            }

            return Task.CompletedTask;
        }


        public override Task InitializingLocal(ITileLocalExtensionContext context)
        {
            ITilePanel panel = context.Workspace.LeftPanel;
            if (panel.Context.ViewContext != null)
            {
                ITile deleteCard = panel.Tiles.TryGet(TileNames.ViewsOther)
                    ?.Tiles.TryGet(DefaultTileNames.DeleteKrDocStateFromView);

                if (deleteCard != null)
                {
                    deleteCard.Context.AddInputBinding(deleteCard, TileKeys.DeleteCardFromView);

                    if (this.session.User.IsAdministrator())
                    {
                        // удаление с зажатым Alt должно быть в режиме "всегда без возможности восстановления",
                        // но виртуальная карточка и так удаляется без возможности восстановления,
                        // поэтому просто добавим горячую клавишу, чтобы диалог отображался и с Alt, и без него
                        deleteCard.Context.AddInputBinding(deleteCard, TileKeys.DeleteCardFromViewAlt);
                    }
                }
            }

            return Task.CompletedTask;
        }

        #endregion
    }
}